content/blog/2016/2016-04-07-pipeline-for-runs-on-hardware.adoc

```diff
@@ -0,0 +1,258 @@
+
+WARNING: Disclaimer. 
+Makefile above is a sample stuff for demo purposes.
+Never write such makefiles.
+
+It is possible to continue expanding the pipeline in such way.
+link:https://github.com/jenkinsci/pipeline-examples[Pipeline Examples]
+contain examples for common cases: build parallelization, code sharing between pipelines, error handling, etc.
+
+## Lessons learned
+
+During the last 2 years I've tried using Pipeline for Hardware test automation several times.
+The first attempts were not very successful, but the ecosystem has been evolving rapidly. 
+There are still several missing integrations, but I feel Pipeline becomes a really powerful tool.
+
+I would like to mention the following improvement areas:
+
+* *Shared resource management across pipelines*. It can be done by the incoming Pipeline integration in the 
+link:https://wiki.jenkins-ci.org/display/JENKINS/Lockable+Resources+Plugin[Lockable Resources plugin] 
+(link:https://issues.jenkins-ci.org/browse/JENKINS-30269[JENKINS-30269]).
+Another case is integration with 
+link:https://wiki.jenkins-ci.org/display/JENKINS/Throttle+Concurrent+Builds+Plugin[Throttle Concurrent Builds plugin], which is an effective engine for limiting the license utilization in automation infrastructures 
+(link:https://issues.jenkins-ci.org/browse/JENKINS-31801[JENKINS-31801]).
+* *Better support of CLI tools*. 
+EDA tools frequently need a complex environment, which should be deployed on nodes somehow. 
+Integration with 
+link:https://wiki.jenkins-ci.org/display/JENKINS/Custom+Tools+Plugin[Custom Tools Plugin] seems to be the best option, especially in the case of multiple tool versions 
+(link:https://issues.jenkins-ci.org/browse/JENKINS-30680[JENKINS-30680]).
+* *Pipeline package manager* with dependency management, which would allow developing Pipeline libraries and sharing them between teams. 
+link:https://github.com/jenkinsci/workflow-plugin/blob/master/cps-global-lib/README.md[Pipeline Global Library] and
+link:https://github.com/jenkinsci/workflow-remote-loader-plugin[Pipeline Remote Loader] can be used as a workaround.
+* *Pipeline debugger*. HW test runs are very slow, so it is difficult to troubleshoot and fix issues in the Pipeline code if you have to run every build from scratch. 
+There are several features in Pipeline, which simplify the development, but we still need a full-featured IDE.
+
+## Conclusions
+
+Jenkins is a powerful *automation framework*, which can be used in many areas.
+Even though Jenkins has no dedicated plugins for test runs on hardware, it provides many general-purpose "building blocks", which allow implementing almost any flow.
+That's why Jenkins is so popular in the hardware and embedded areas.
+
+Pipeline-as-code is an emerging technology, which should greatly simplify the implementation of complex flows.
+Currently it lacks integrations with particular Jenkins features, but hopefully this issue will be solved soon.
```

Comment:
This is a great way to make something that was being made positive sound very negative.

I know what is being said here, but a more positive phrasing would be something like:

> Pipeline continues to evolve and improve but if you're developing embedded projects, consider Pipeline as a durable, extensible and versatile means of implementing your automation.

(This would naturally subsume the sentence on line 223)